File: GGAndroid_2.0/GGOpen/open-twitter/src/main/java/io/ganguo/twitter/share/TwitterShareService.java
package io.ganguo.twitter.share;

import android.content.Context;
import android.net.Uri;
import android.os.Build;
import android.support.v4.content.FileProvider;

import com.twitter.sdk.android.tweetcomposer.TweetComposer;

import java.io.File;
import java.net.MalformedURLException;
import java.net.URL;

import io.ganguo.open.sdk.IService;
import io.ganguo.twitter.BuildConfig;
import io.ganguo.twitter.entity.TwitterShareData;
import io.ganguo.utils.util.Files;
import io.ganguo.utils.util.Strings;

/**
 * <p>
 * Twitter Share Service
 * </p>
 * Created by leo on 2018/9/30.
 */
public class TwitterShareService implements IService {
    private TwitterShareData shareData;
    private Context context;

    private TwitterShareService() {
    }


    /**
     * 单例
     *
     * @return
     */
    public static TwitterShareService get() {
        return LazyHolder.HOLDER;
    }


    /**
     * 单例 - 懒加载
     */
    public static class LazyHolder {
        private static TwitterShareService HOLDER = new TwitterShareService();
    }

    /**
     * function: 设置分享参数
     *
     * @param context
     * @param data
     * @return
     */
    public TwitterShareService applyShareData(Context context, TwitterShareData data) {
        this.shareData = data;
        this.context = context;
        return this;
    }

    /**
     * function:share twitter
     *
     * @return
     */
    @Override
    public boolean apply() {
        return share();
    }


    /**
     * function:share twitter
     *
     * @return
     */
    protected boolean share() {
        TweetComposer.Builder builder = null;
        try {
            builder = new TweetComposer
                    .Builder(context)
                    .text(shareData.getText())
                    .url(new URL(shareData.getUrl()));
            if (shareData.getFileUri() != null) {
                builder.image(shareData.getFileUri());
            }
        } catch (MalformedURLException e) {
            e.printStackTrace();
        }
        if (builder != null) {
            builder.show();
            return true;
        }
        return false;
    }


    /**
     * function:资源释放
     *
     * @return
     */
    @Override
    public void release() {
        shareData = null;
        context = null;
    }
}

File: GGAndroid_3.0/core/service/open/open-twitter/src/main/java/io/ganguo/twitter/share/TwitterShareService.kt
package io.ganguo.twitter.share

import android.app.Activity
import com.twitter.sdk.android.tweetcomposer.TweetComposer
import io.ganguo.open.sdk.ShareResultObservable
import io.ganguo.open.sdk.annotation.OpenChannel
import io.ganguo.open.sdk.annotation.OpenStatus
import io.ganguo.open.sdk.entity.OpenResult
import io.ganguo.twitter.ATwitterShareService
import io.ganguo.twitter.TwitterHandler
import io.ganguo.twitter.entity.TwitterShareEntity
import io.reactivex.Observable
import java.lang.ref.WeakReference
import java.net.URL

/**
 * <pre>
 *     author : leo
 *     time   : 2019/11/05
 *     desc   : Twitter 分享服务
 * </pre>
 */
class TwitterShareService(activity: Activity, private var shareEntity: TwitterShareEntity, var listener: TwitterShareListener) : ATwitterShareService() {
    private var weakActivity: WeakReference<Activity>? = WeakReference(activity)


    /**
     * 调用推特分享
     * @return ShareResultObservable
     */
    override fun startService(): ShareResultObservable {
        try {
            buildTweetBuilder().show()
            listener.onComplete()
        } catch (e: Throwable) {
            e.printStackTrace()
            return Observable.just(OpenResult(OpenChannel.TWITTER, OpenStatus.FAILED, e.message))
        }
        return resultObserver
    }

    /**
     * 检测推特SDK配置是否正常
     * @return Throwable?
     */
    override fun checkException(): Throwable? {
        return TwitterHandler.checkException()
    }


    /**
     * 创建分享数据对象
     * @return TweetComposer.Builder
     */
    private fun buildTweetBuilder(): TweetComposer.Builder {
        var builder = TweetComposer.Builder(weakActivity?.get()!!)
                .text(shareEntity.text);
        if (shareEntity.webPageUrl.isNotEmpty()) {
            builder.url(URL(shareEntity.webPageUrl))
        }
        if (shareEntity.imageFileUri != null) {
            builder!!.image(shareEntity.imageFileUri)
        }
        return builder
    }

    /**
     * 资源释放
     */
    override fun release() {
        super.release()
        weakActivity?.clear()
        weakActivity = null
    }
}